Enable mypy allow redefinition

ringohoffman opened the pull request:
Related #101528

I tried to enable this in another PR but it uncovered a bunch of type errors: https://github.com/pytorch/pytorch/actions/runs/4999748262/jobs/8956555243?pr=101528#step:10:1305

The goal of this PR is to fix these errors.

---

This PR enables [allow_redefinition = True](https://mypy.readthedocs.io/en/stable/config_file.html#confval-allow_redefinition) in `mypy.ini`, which allows for a common pattern:

> Allows variables to be redefined with an arbitrary type, as long as the redefinition is in the same block and nesting level as the original definition.

`allow_redefinition` allows mypy to be more flexible by allowing reassignment to an existing variable with a different type... for instance (from the linked PR):

https://github.com/pytorch/pytorch/blob/4a1e9230bae52aba3d0d03dc13f0416f464b2679/torch/nn/parallel/data_parallel.py#L213

A `Sequence[Union[int, torch.device]]` is narrowed to `Sequence[int]` thru reassignment to the same variable.

cc: @ezyang @albanD

Commit:
Enable allow_redefinition

This allows for a common pattern, which is type narrowing by reassigning to a variable within the same scope: Allows variables to be redefined with an arbitrary type, as long as the redefinition is in the same block and nesting level as the original definition. See: https://mypy.readthedocs.io/en/stable/config_file.html#confval-allow_redefinition

Commit:
Fix mypy var-annotated error

If the TypeVar does not get bound in the call to the constructor, mypy cannot determine the type of the generic return value

Commit:
Fix mypy union-attr error

Remove the type declaration since mypy will not overwrite the declared type, so even after assigning it to a Tensor, mypy will infer the declared variable as Optional

Commit:
Fix mypy operator error

Unsupported addition for tuple and list, just make the value a list

Commit:
Add and improve type hints

Replace Module with GraphModule where appropriate, add return types

Commit:
Replace torch.round with Tensor.item + builtins.round

This way we can ensure n is an int to appease torch.Size constructor typing

Commit:
Assign result of os.getenv to a variable

If we assign the result to a variable, we can narrow the variable's type and make sure cuda_flags is a list[str]

Commit:
Avoid using norm_type as both an int and str

We can avoid a mypy assignment error if we use a different name for int(norm_type) instead of reassigning it to norm_type

Commit:
Assert node.target is callable if node.op == 'call_function'

Fixes a mypy arg-type error since node.target can be a str, but not if node.op == 'call_function' (based on the Node constructor logic)

Commit:
Use list comprehension instead of map

This also fixes the mypy operator error from adding a tuple and a list

Skylion007 commented on mypy.ini:
```diff
@@ -5,6 +5,7 @@
 plugins = mypy_plugins/check_mypy_version.py, numpy.typing.mypy_plugin
 
```
mypy-strict and mypy-nofollow are two separate inis. Don't you need to edit those too?

ringohoffman replied:
I wasn't aware of those files. I'm sure we would want to be consistent across them so I'll add it.

Commit:
Add allow_redefinition=True

Allows variables to be redefined with an arbitrary type, as long as the redefinition is in the same block and nesting level as the original definition.

Commit:
Avoid making args a list

We reuse the pattern used for native_batch_norm s.t. we keep args a Tuple[Any, ...]; minor optimization to use tuple instead of list in membership checks, readability improvement in using comprehension instead of map; import torch.fx since fx is not a known member of torch

Commit:
Format for line length limit

ringohoffman commented:
From: https://hud.pytorch.org/pr/pytorch/pytorch/102046#13684112327
> 2023-05-23T09:15:18.1963963Z An FX function that has been marked as backwards-compatible has experienced a signature change. See the above exception context for more information. If this change was unintended, please revert it. If it was intended, check with the FX team to ensure that the proper deprecation protocols have been followed and subsequently --accept the change.
Which seems to be coming from this change: https://github.com/pytorch/pytorch/pull/102046/files#diff-15d350059f273cfb14a94fe2a098ee60e59ae333719bbe428420a1c4c20ebcbdR353

@ezyang is this an acceptable change?

Commit:
Remove unnecessary improvement

This is an improvement but not strictly necessary to fix the type errors we were seeing, and a change to the signature of this method needs specific approval, so we can just leave it alone for now

Skylion007 commented on torch/_decomp/decompositions.py:
```diff
@@ -1226,9 +1226,9 @@ def native_group_norm_backward(
-    d_input: Optional[Tensor] = None
-    d_gamma: Optional[Tensor] = None
-    d_bias: Optional[Tensor] = None
+    d_input = None
```
Why does removing these types help? The typing should be fixed, not removed.

ringohoffman replied:
For this specific line, with `allow_redefinition=True`: https://github.com/pytorch/pytorch/actions/runs/4999748262/jobs/8956555243?pr=101528#step:10:1326


```python
d_input = (
    torch.mul(grad_output.reshape(N, group, cpg, HxW), c1)
    + torch.mul(input.reshape(N, group, cpg, HxW), c2)
    + c3
reveal_type(d_input)  # note: Revealed type is "Union[torch._tensor.Tensor, None]"
)
```
mypy thinks the result of this is somehow `Optional[Tensor]`.

With `allow_redefinition=False`, the revealed type is `"torch._tensor.Tensor"` as expected, so I suppose this is a bug with `allow_redefinition` stemming from how it is able to infer types from the above expression?

I would be open to using a `type: ignore[union-attr]` here instead of removing the type declarations.

ringohoffman replied:
Actually I found the source of the error 1f810e0ac1b06a91b4f6fa11ca529d22c65d37f6 😵‍💫 really obscure

Commit:
Add return type to _unsqueeze_to_dim

This was the actual apparent source of the union-attr error; because there was no return type on that function, c2 and c3 were being inferred as Any in allow_redefinition, leading to d_input not being narrowed to Tensor and remaining Optional[Tensor]

ezyang commented:
@pytorchbot merge
